### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [131]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [132]:
# Initialise data to lists
total = len(purchase_data.index)
data = [
    {
        'Total Players': total
    }
] 

# Create DataFrame
df = pd.DataFrame(data) 
  
# Print the data 
df 

Total Players
0            780

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [133]:
purchase_data.describe()
items = purchase_data["Item ID"].unique()

num_unique_items = len(items)
avg_price = purchase_data["Price"].mean()
num_purchases = len(purchase_data["Item ID"])
total_revenue = purchase_data["Price"].sum()

# Initialise data to lists
data = [
    {
        'Number of Unique Items': num_unique_items, 
        'Average Price': avg_price, 
        'Number of Purchases': num_purchases, 
        'Total Revenue': total_revenue
    }
] 
  
# Create DataFrame
df = pd.DataFrame(data) 

# Format
df["Average Price"] = df["Average Price"].map("${0:,.2f}".format)
df["Total Revenue"] = df["Total Revenue"].map("${0:,.2f}".format)
  
# Print the data 
df 

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [134]:
players = purchase_data["Gender"].value_counts()
num_players = players.sum()

# Initialise data to lists
data = [
    {
        'Total Count': players["Male"], 
        'Percentage of Players': players["Male"] / num_players * 100
    },
    {
        'Total Count': players["Female"], 
        'Percentage of Players': players["Female"] / num_players * 100
    },
    {
        'Total Count': players["Other / Non-Disclosed"], 
        'Percentage of Players': players["Other / Non-Disclosed"] / num_players * 100
    }
] 
  
# Create DataFrame
df = pd.DataFrame(data, index = ['Male', 'Female', 'Other / Non-Disclosed'])

# Format
df["Percentage of Players"] = df["Percentage of Players"].map("{0:,.2f}%".format)

# Print the data 
df 

Total Count Percentage of Players
Male                           652                83.59%
Female                         113                14.49%
Other / Non-Disclosed           15                 1.92%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [144]:
gender_df = purchase_data.groupby(["Gender"])

# Find purchase count for each gender
purchase_counts = gender_df["Purchase ID"].count()

# Find average count for each gender
avg_purchase = gender_df["Price"].mean()

# Find total purchase value for each gender
total_purchase = gender_df["Price"].sum()

# Find average purchase total per person by gender
avg_purchase_total = total_purchase / gender_count
avg_purchase_total

# Create DataFrame
purchase_df = pd.DataFrame(
    {
        "Purchase Count": purchase_counts,
        "Average Purchase Price": avg_purchase,
        "Total Purchase Value": total_purchase,
        "Average Total Purchase per Person": avg_purchase_total
    }
)

# Format
purchase_df["Average Purchase Price"] = purchase_df["Average Purchase Price"].map("${0:,.2f}".format)
purchase_df["Average Total Purchase per Person"] = purchase_df["Average Total Purchase per Person"].map("${0:,.2f}".format)
purchase_df["Total Purchase Value"] = purchase_df["Total Purchase Value"].map("${0:,.2f}".format)

purchase_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Average Total Purchase per Person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [136]:
min_age = purchase_data["Age"].min() # 7
max_age = purchase_data["Age"].max() # 45

bins = [7, 9.9, 14.9, 19.9 , 24.9, 29.9, 34.9, 39.9, 45]
label_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Append column age range
purchase_data["Age Ranges"] = pd.cut(purchase_data["Age"], bins, labels = label_names, include_lowest = True)

# Group df by age ranges
age_df = purchase_data.groupby(["Age Ranges"])

# Count of unique SN
age_count = age_df["SN"].nunique()

# Calculate percentages by age group
age_percentage = age_count / total * 100

# Create summary DataFrame
age_demographic_df = pd.DataFrame(
    {
        "Total Count": age_count,
        "Percentage of Players": age_percentage
    }
)

# Format
age_demographic_df = age_demographic_df.rename_axis([""])
age_demographic_df["Percentage of Players"] = age_demographic_df["Percentage of Players"].map("{0:,.2f}%".format)

age_demographic_df

Total Count Percentage of Players
                                        
<10             17                 2.18%
10-14           22                 2.82%
15-19          107                13.72%
20-24          258                33.08%
25-29           77                 9.87%
30-34           52                 6.67%
35-39           31                 3.97%
40+             12                 1.54%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [139]:
purchase_count = age_df["Purchase ID"].count()
avg_purchase = age_df["Price"].mean()
total_value = age_df["Price"].sum()

avg_total = total_value / age_count

# Create summary DataFrame
purchasing_analysis_df = pd.DataFrame (
    {
        "Purchase Count": purchase_count,
        "Average Purchase Price": avg_purchase,
        "Total Purchase Value": total_value,
        "Average Purchase Total per Person": avg_total
    }
)

# Format
purchasing_analysis_df["Average Purchase Price"] = purchasing_analysis_df["Average Purchase Price"].map("${0:,.2f}".format)
purchasing_analysis_df["Total Purchase Value"] = purchasing_analysis_df["Total Purchase Value"].map("${0:,.2f}".format)
purchasing_analysis_df["Average Purchase Total per Person"] = purchasing_analysis_df["Average Purchase Total per Person"].map("${0:,.2f}".format)

purchasing_analysis_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Average Purchase Total per Person  
Age Ranges                                    
<10                                    $4.54  
10-14                                  $3.76  
15-19                                  $3.86  
20-24                                  $4.32  
25-29                                  $3.81  
30-34                                  $4.12  
35-39                                  $4.76  
40+                                    $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [142]:
spenders_df = purchase_data.groupby(["SN"])

# Calculate total purchases for each SN
spenders_purchase = spenders_df["Price"].sum()

# Calculate purchase count for each spender
purchase_count = spenders_df["Purchase ID"].count()

# Calculate average purchase for each spender
avg_purchase_price = spenders_df["Price"].mean()

# Create summary DataFrame
spenders_data = pd.DataFrame(
    {
        "Purchase Count": purchase_count,
        "Average Purchase Price": avg_purchase_price,
        "Total Purchase Value": spenders_purchase
    }
)

# Sort
top_spenders = spenders_data.sort_values(["Total Purchase Value"], ascending = False)

# Format
top_spenders["Average Purchase Price"] = top_spenders["Average Purchase Price"].map("${0:,.2f}".format) 
top_spenders["Total Purchase Value"] = top_spenders["Total Purchase Value"].map("${0:,.2f}".format) 

top_spenders.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [146]:
popular_items = purchase_data.groupby(["Item ID","Item Name"])

# Calculate purchase count for each purchase ID
purchase_count = popular_items["Purchase ID"].count()

# Calculate total price for each group
total_value = popular_items["Price"].sum()

item_price = total_value / purchase_count

# Create summary DataFrame
top_items = pd.DataFrame(
    {
        "Purchase Count": purchase_count,
        "Item Price": item_price,
        "Total Purchase Value": total_value
    }
)

# Sort
popular_df = top_items.sort_values(["Purchase Count"], ascending = False)

# Format
popular_df["Item Price"] = popular_df["Item Price"].map("${0:,.2f}".format)
popular_df["Total Purchase Value"] = popular_df["Total Purchase Value"].map("${0:,.2f}".format)
                                                          
popular_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [149]:
# Sort df by total purchase value 
profitable_items = top_items.sort_values(["Total Purchase Value"], ascending = False)

# Format
profitable_items["Item Price"] = profitable_items["Item Price"].map("${0:,.2f}".format)
profitable_items["Total Purchase Value"] = profitable_items["Total Purchase Value"].map("${0:,.2f}".format)

profitable_items.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Analysis

Mostly ages 20 to 24 play Heros of Pymoli. This age range comprises 33.08% of the player base.

The majority of players are male. 83.59% are male compared to 14.49% female.

Final Critic is the most popular item and is the most profitable item in the game.